Add text color filter for inactive windows

lilyspiniolas opened the pull request:
#### 46b45c86813860a99b3d6abf585b387ba3414a63
<pre>
Add text color filter for inactive windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=289364">https://bugs.webkit.org/show_bug.cgi?id=289364</a>
<a href="https://rdar.apple.com/146511450">rdar://146511450</a>

Reviewed by Aditya Keerthi, Wenson Hsieh, and Abrar Rahman Protyasha.

Added a new rare inherited RenderStyle flag with the purpose of
filtering text color during the computation of text paint styles.
`TextPaintStyle::computeTextPaintStyle` now checks for the flag
and inverts the text color if the window is inactive and
light-appearance is used. Additionally, the preexisting flag
`insideDefaultButton` has been moved to rare inherited data since
it is not commonly used.

* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::rareInheritedDataChangeRequiresRepaint):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::InheritedFlags::dumpDifferences const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::insideDefaultButton const):
(WebCore::RenderStyle::shouldApplyColorFilterWhenInactive const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setInsideDefaultButton):
(WebCore::RenderStyle::setShouldApplyColorFilterWhenInactive):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/291934@main">https://commits.webkit.org/291934@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5fd4492dcc21c24d4be4498cf02ecf9a9c2d53b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94448 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14040 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3820 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99467 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44975 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96498 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14336 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22468 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/72080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29402 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97450 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10659 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/85277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52410 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2969 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44290 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/3070 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101511 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21502 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21751 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81297 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80455 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25000 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14727 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15158 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21479 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/26633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21163 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24623 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22904 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2440,6 +2442,7 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
         unsigned char printColorAdjust : 1; // PrintColorAdjust
         unsigned char insideLink : 2; // InsideLink
         unsigned char insideDefaultButton : 1;
+        unsigned char shouldApplyColorFilterWhenInactive : 1;
```
Do we need to update the comment below? (`// Total = 57 bits (fits in 8 bytes)`)

(Also the `Color Stuff = 5 bits` comment above)

lilyspiniolas replied:
good catch, fixing now

whsieh reviewed (approved):
LGTM with a minor comment

pxlcoder commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2356,6 +2352,14 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
     inline OptionSet<PositionVisibility> positionVisibility() const;
     inline void setPositionVisibility(OptionSet<PositionVisibility>);
 
+    static bool initialInsideDefaultButton() { return false; }
```
Can remove this.

pxlcoder commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -2356,6 +2352,14 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
+    inline bool insideDefaultButton() const;
+    inline void setInsideDefaultButton(bool);
+
+    static bool initialShouldApplyColorFilterWhenInactive() { return false; }
```
Can remove this.

pxlcoder commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1343,8 +1342,7 @@ bool RenderStyle::changeRequiresRepaint(const RenderStyle& other, OptionSet<Styl
     if (m_inheritedFlags.visibility != other.m_inheritedFlags.visibility
         || m_inheritedFlags.printColorAdjust != other.m_inheritedFlags.printColorAdjust
-        || m_inheritedFlags.insideLink != other.m_inheritedFlags.insideLink
-        || m_inheritedFlags.insideDefaultButton != other.m_inheritedFlags.insideDefaultButton)
```
You'll need to update `rareInheritedDataChangeRequiresRepaint` to check for `insideDefaultButton` now that it is no longer checked here.

pxlcoder commented on Source/WebCore/rendering/TextPaintStyle.cpp:
```diff
@@ -104,6 +104,16 @@ TextPaintStyle computeTextPaintStyle(const RenderText& renderer, const RenderSty
+        auto usingDarkAppearance = renderer.styleColorOptions().contains(StyleColorOptions::UseDarkAppearance);
+        RefPtr page = renderer.frame().page();
+        if (page && !usingDarkAppearance && !page->focusController().isActive()) {
+            const auto color = paintStyle.fillColor;
```
Can remove `const`, it's not commonly used in WebKit.

lilyspiniolas committed:
Add text color filter for inactive windows
https://bugs.webkit.org/show_bug.cgi?id=289364
rdar://146511450

Reviewed by Aditya Keerthi, Wenson Hsieh, and Abrar Rahman Protyasha.

Added a new rare inherited RenderStyle flag with the purpose of
filtering text color during the computation of text paint styles.
`TextPaintStyle::computeTextPaintStyle` now checks for the flag
and inverts the text color if the window is inactive and
light-appearance is used. Additionally, the preexisting flag
`insideDefaultButton` has been moved to rare inherited data since
it is not commonly used.

* Source/WebCore/rendering/TextPaintStyle.cpp:
(WebCore::computeTextPaintStyle):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::rareInheritedDataChangeRequiresRepaint):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::InheritedFlags::dumpDifferences const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::insideDefaultButton const):
(WebCore::RenderStyle::shouldApplyColorFilterWhenInactive const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setInsideDefaultButton):
(WebCore::RenderStyle::setShouldApplyColorFilterWhenInactive):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: https://commits.webkit.org/291934@main
